File: problematic_code_test/vulnerable_code.js
/**
 * JavaScript file with security vulnerabilities for testing
 */

// SQL Injection vulnerability
function getUserData(userId) {
  // Never concatenate user input into SQL queries
  const query = `SELECT * FROM users WHERE id = ${userId}`;
  // db.query(query); // Would execute unsafe query
  return query;
}

// Command injection vulnerability
function runCommand(filename) {
  const { exec } = require('child_process');
  // Never pass user input directly to exec
  exec(`cat ${filename}`, (error, stdout, stderr) => {
    console.log(stdout);
  });
}

// Hardcoded credentials
const config = {
  databasePassword: 'admin123',  // Never hardcode passwords
  apiKey: 'sk-1234567890abcdefghijklmnop',  // Never hardcode API keys
  secretToken: 'secret_token_12345',  // Never hardcode tokens
  awsAccessKey: 'AKIAIOSFODNN7EXAMPLE',  // AWS credentials
  awsSecretKey: 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY'
};

// Eval with user input (code injection)
function evaluateExpression(expr) {
  // eval executes arbitrary JavaScript code
  return eval(expr);
}

// Unsafe regex (ReDoS vulnerability)
function validateEmail(email) {
  // Catastrophic backtracking possible
  const regex = /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/;
  return regex.test(email);
}

// XSS vulnerability - direct HTML injection
function displayUserContent(userInput) {
  // Never insert user input directly into HTML
  document.getElementById('content').innerHTML = userInput;
}

// Insecure random number generation
function generateToken() {
  // Math.random() is not cryptographically secure
  return Math.floor(Math.random() * 1000000);
}

// Path traversal vulnerability
function readFile(filename) {
  const fs = require('fs');
  // No validation allows ../../etc/passwd
  const path = `/var/data/${filename}`;
  return fs.readFileSync(path, 'utf8');
}

// Prototype pollution vulnerability
function mergeObjects(target, source) {
  // Vulnerable to prototype pollution
  for (const key in source) {
    if (typeof source[key] === 'object') {
      target[key] = mergeObjects(target[key] || {}, source[key]);
    } else {
      target[key] = source[key];
    }
  }
  return target;
}

// Using deprecated/vulnerable functions
function hashPassword(password) {
  const crypto = require('crypto');
  // MD5 is cryptographically broken
  return crypto.createHash('md5').update(password).digest('hex');
}

// Open redirect vulnerability
function redirectUser(url) {
  // No validation allows redirect to malicious sites
  window.location.href = url;
}

// localStorage with sensitive data
function storeSensitiveData(creditCard, ssn) {
  // Never store sensitive data in localStorage
  localStorage.setItem('creditCard', creditCard);
  localStorage.setItem('ssn', ssn);
}

// CORS misconfiguration
function setupCORS(req, res) {
  // Allows any origin - security risk
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Credentials', 'true');
}

// Timing attack vulnerability
function comparePasswords(input, actual) {
  // Early return creates timing differences
  if (input.length !== actual.length) {
    return false;
  }
  
  for (let i = 0; i < input.length; i++) {
    if (input[i] !== actual[i]) {
      return false;  // Early return leaks information
    }
  }
  return true;
}

// Unsafe deserialization
function deserializeData(data) {
  // Function constructor can execute arbitrary code
  return new Function('return ' + data)();
}

// Information disclosure in error messages
function loginUser(username, password) {
  if (username === 'admin' && password !== 'correct') {
    // Reveals that 'admin' is a valid username
    throw new Error('Invalid password for admin user');
  }
}

// Client-side authentication
function checkAuth() {
  // Never do authentication on client side only
  const isAdmin = document.cookie.includes('admin=true');
  if (isAdmin) {
    showAdminPanel();
  }
}

// Weak session management
function createSession(userId) {
  // Predictable session ID
  const sessionId = userId + '_' + Date.now();
  document.cookie = `session=${sessionId}`;
}

// NoSQL injection
function findUser(username) {
  // Vulnerable to NoSQL injection with MongoDB
  const query = { username: username };
  // db.users.find(query);
  return query;
}

// XXE in XML parsing
function parseXML(xmlString) {
  const parser = new DOMParser();
  // Default parser may be vulnerable to XXE
  const xmlDoc = parser.parseFromString(xmlString, "text/xml");
  return xmlDoc;
}

// Export for testing
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    getUserData,
    evaluateExpression,
    generateToken,
    config
  };
}

File: problematic_code_test/vulnerable_code.py
"""Python file with security vulnerabilities for ASH and CodeQL testing."""

import os
import pickle
import subprocess
import hashlib
import random
from typing import Any


# SQL Injection vulnerability
def get_user_data(user_id: str):
    """Vulnerable to SQL injection."""
    # Never use string formatting for SQL queries
    query = f"SELECT * FROM users WHERE id = {user_id}"
    # In real code, would execute: cursor.execute(query)
    return query


# Command injection vulnerability
def run_system_command(filename: str):
    """Vulnerable to command injection."""
    # Never use shell=True with user input
    result = subprocess.run(
        f"cat {filename}",
        shell=True,
        capture_output=True,
        text=True
    )
    return result.stdout


# Hardcoded credentials (security issue)
def connect_to_database():
    """Contains hardcoded credentials."""
    DATABASE_PASSWORD = "admin123"  # Never hardcode passwords
    API_KEY = "sk-1234567890abcdefghijklmnop"  # Never hardcode API keys
    SECRET_TOKEN = "secret_token_12345"  # Never hardcode tokens
    
    connection_string = f"postgresql://admin:{DATABASE_PASSWORD}@localhost/db"
    return connection_string


# Unsafe deserialization
def load_user_data(serialized_data: bytes) -> Any:
    """Vulnerable to unsafe deserialization attacks."""
    # pickle.loads can execute arbitrary code
    return pickle.loads(serialized_data)


# Path traversal vulnerability
def read_file(filename: str):
    """Vulnerable to path traversal attacks."""
    # No input validation allows ../../../etc/passwd
    filepath = f"/var/data/{filename}"
    with open(filepath, 'r') as f:
        return f.read()


# Weak cryptography
def hash_password(password: str):
    """Uses weak hashing algorithm."""
    # MD5 is cryptographically broken, never use for passwords
    return hashlib.md5(password.encode()).hexdigest()


# Insecure random number generation
def generate_token():
    """Uses insecure random for security-sensitive operation."""
    # random module is not cryptographically secure
    token = random.randint(100000, 999999)
    return str(token)


# Eval with user input (code injection)
def calculate_expression(expr: str):
    """Vulnerable to code injection via eval."""
    # eval() executes arbitrary Python code
    result = eval(expr)
    return result


# XXE vulnerability in XML parsing
def parse_xml(xml_string: str):
    """Vulnerable to XML External Entity attacks."""
    import xml.etree.ElementTree as ET
    # Default XML parser is vulnerable to XXE
    root = ET.fromstring(xml_string)
    return root


# SSRF vulnerability
def fetch_url(url: str):
    """Vulnerable to Server-Side Request Forgery."""
    import urllib.request
    # No validation allows internal network access
    response = urllib.request.urlopen(url)
    return response.read()


# Insecure temporary file
def create_temp_file(data: str):
    """Creates predictable temporary file."""
    # Predictable filename in shared directory
    temp_file = f"/tmp/app_temp_{os.getpid()}.txt"
    with open(temp_file, 'w') as f:
        f.write(data)
    return temp_file


# Using assert for security (gets removed in optimized code)
def validate_admin(user_role: str):
    """Incorrectly uses assert for security check."""
    # assert statements are removed with python -O
    assert user_role == "admin", "Unauthorized access"
    return True


# Logging sensitive information
def process_payment(card_number: str, cvv: str):
    """Logs sensitive payment information."""
    import logging
    # Never log sensitive data
    logging.info(f"Processing payment for card: {card_number}, CVV: {cvv}")
    return "Payment processed"


# Race condition vulnerability
COUNTER = 0

def increment_counter():
    """Has race condition in multi-threaded environment."""
    global COUNTER
    # Not thread-safe, can lead to race conditions
    temp = COUNTER
    temp = temp + 1
    COUNTER = temp
    return COUNTER


# Open redirect vulnerability
def redirect_user(next_url: str):
    """Vulnerable to open redirect attacks."""
    # No validation allows redirect to malicious sites
    return f"Location: {next_url}"


if __name__ == "__main__":
    # Example of unsafe operations (DO NOT RUN)
    print("This file contains intentional security vulnerabilities")
    print("For testing ASH and CodeQL security scanning only")
    
    # These would be flagged by security scanners
    password = "weak_password"
    weak_hash = hash_password(password)
    print(f"Weak hash: {weak_hash}")
    
    # Hardcoded secret in string
    secret = "aws_secret_access_key_AKIAIOSFODNN7EXAMPLE"